server/src/test/java/io/spine/examples/shareaware/server/e2e/given/WalletBalanceSubscription.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2023, TeamDev. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.examples.shareaware.server.e2e.given;
+
+import io.spine.client.Client;
+import io.spine.core.UserId;
+import io.spine.examples.shareaware.wallet.WalletBalance;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+import static io.spine.util.Exceptions.*;
+
+/**
+ * Subscription for the {@code WalletBalance} state changes.
+ */
+public final class WalletBalanceSubscription {
```

Comment:
How about a cleaner approach?

```java
/**
 * Subscription for the {@code WalletBalance} state changes.
 */
public final class WalletBalanceSubscription {

    private final List<CompletableFuture<WalletBalance>> balances = new ArrayList<>();
    
    private final ObservedBalance balance = new ObservedBalance();

    WalletBalanceSubscription(Client client, UserId user) {
        client.onBehalfOf(user)
              .subscribeTo(WalletBalance.class)
              .observe(balance::set)
              .post();
    }

    /**
     * Provides current state of the {@code WalletBalance}.
     */
    public WalletBalance balance()  {
        return balance.value();
    }
    
    private static final class ObservedBalance {
        
        private CompletableFuture<WalletBalance> future = new CompletableFuture<>();
        
        private void set(WalletBalance value) {
            if(future.isDone()) {
                future = new CompletableFuture<>();
            }
            future.complete(value);
        }
        
        private WalletBalance value() {
            try {
                return future.get();
            } catch (InterruptedException | ExecutionException e) {
                throw illegalStateWithCauseOf(e);
            }
        }
    }
}
```

It's a quicky, just to illustrate the idea.